lib/blocks.php

```diff
@@ -179,6 +179,34 @@ function gutenberg_register_core_block_styles( $block_name ) {
 
 		// Add a reference to the stylesheet's path to allow calculations for inlining styles in `wp_head`.
 		wp_style_add_data( "wp-block-{$block_name}", 'path', gutenberg_dir_path() . $style_path );
+
+		// If the current theme supports wp-block-styles, dequeue the full stylesheet
+		// and instead attach each block's theme-styles to their block styles stylesheet.
+		if ( current_theme_supports( 'wp-block-styles' ) ) {
+
+			// Dequeue the full stylesheet.
+			// Make sure this only runs once, it doesn't need to run for every block.
+			static $stylesheet_removed;
+			if ( ! $stylesheet_removed ) {
+				add_action(
+					'wp_enqueue_scripts',
+					function() {
+						wp_dequeue_style( 'wp-block-library-theme' );
+					}
+				);
+				$stylesheet_removed = true;
+			}
+
+			// Get the path to the block's stylesheet.
+			$theme_style_path = is_rtl()
+				? gutenberg_dir_path() . "build/block-library/blocks/$block_name/theme-rtl.css"
+				: gutenberg_dir_path() . "build/block-library/blocks/$block_name/theme.css";
+
+			// If the file exist, add its contents to the main block styles.
+			if ( file_exists( $theme_style_path ) ) {
```

Comment:
Could be possible not having `style.css` and have `theme.css`? Because this runs only if `style.css` exists.

Reply:
Thanks for the flag! Fixed :+1: